nats/replica_client.go

```diff
@@ -0,0 +1,403 @@
+package nats
+
+import (
+	"context"
+	"fmt"
+	"io"
+	"strings"
+	"time"
+
+	"github.com/benbjohnson/litestream"
+	"github.com/benbjohnson/litestream/internal"
+	"github.com/nats-io/nats.go"
+	"github.com/nats-io/nats.go/jetstream"
+	"github.com/valyala/bytebufferpool"
+)
+
+const (
+	ReplicaClientType     = "nats"
+	DefaultBucketName     = "litestream"
+	DefaultBucketReplicas = 3
+	DefaultBucketMaxBytes = 0               // unlimited
+	DefaultBucketTTL      = 0 * time.Second // unlimited
+
+	generationsPrefix = "generations/"
+)
+
+type ReplicaClient struct {
+	nc          *nats.Conn
+	objectStore jetstream.ObjectStore
+
+	// NATS connection info
+	BucketName     string
+	BucketReplicas int
+	BucketMaxBytes int64
+	BucketTTL      time.Duration
+}
+
+func NewReplicaClient(nc *nats.Conn) *ReplicaClient {
+	return &ReplicaClient{
+		nc:             nc,
+		BucketName:     DefaultBucketName,
+		BucketReplicas: DefaultBucketReplicas,
+		BucketMaxBytes: DefaultBucketMaxBytes,
+		BucketTTL:      DefaultBucketTTL,
+	}
+}
+
+// Type returns the type of client.
+func (c *ReplicaClient) Type() string {
+	return ReplicaClientType
+}
+
+// Returns a list of available generations. Order is undefined.
+func (c *ReplicaClient) Generations(ctx context.Context) (_ []string, err error) {
```

Comment:
I did not peak at the other backend implementations, so if this is the idiomatic expectation in the codebase then 👍. However, at least for this backend implementation is a bit more difficult to read which values are being returned at times across some of the methods since there is scoped explicit assignment of `err` in particular vs. implicit.

Reply:
Sorry, I didn't define what "it" is in the first sentence.. named return values.